From 50c5ad120aa8d92ba60a856ccadfa69254e75ae9 Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Fri, 3 Nov 2023 17:39:22 +0100
Subject: [PATCH 0059/1695] plat-stm32mp1: syscfg: STM32MP13 dynamic IO
 compensation

Replace IO compensation API functions
stm32mp_syscfg_enable_io_compensation() and
stm32mp_syscfg_disable_io_compensation() with a new API function
stm32mp_set_io_comp_by_index() dedicated to runtime configuration
of STM32MP13 SDMMC's domains IO compensation only.

On STM32MP15 variant, the configuration is enabled only during
initialization. On STM32MP13 variant, the same feature is also enabled
during initialization but the device embeds 2 more IO domains
(SDMMC1 and SDMMC2) for which the new API function allow runtime
reconfiguration support.

For sake of simplicity, keep related clocks always on.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 79d1d090995303b9081f759f0571fb4beb0189f1 plat-stm32mp1: syscfg: STM32MP13 dynamic IO compensation]
Change-Id: I30aba6f9615d5dd66456d721e3df1cf00adfe1f3
Co-developed-by: Lionel Debieve <lionel.debieve@foss.st.com>
Signed-off-by: Lionel Debieve <lionel.debieve@foss.st.com>
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
[etienne: cherry-picked from 5611e846a0d62acd376e6891c9caf5530b5ef9ad]
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
---
 .../plat-stm32mp1/drivers/stm32mp1_syscfg.c   | 99 ++++++++++++-------
 .../plat-stm32mp1/drivers/stm32mp1_syscfg.h   | 24 +++++
 core/arch/arm/plat-stm32mp1/stm32_util.h      |  7 --
 3 files changed, 89 insertions(+), 41 deletions(-)
 create mode 100644 core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.h

diff --git a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.c b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.c
index 208c1e835..8f795caf3 100644
--- a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.c
+++ b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.c
@@ -3,8 +3,10 @@
  * Copyright (c) 2019-2023, STMicroelectronics
  */
 
+#include <config.h>
 #include <drivers/clk.h>
 #include <drivers/stm32mp_dt_bindings.h>
+#include <drivers/stm32mp1_syscfg.h>
 #include <initcall.h>
 #include <kernel/delay.h>
 #include <mm/core_memprot.h>
@@ -18,6 +20,8 @@
  */
 #define SYSCFG_CMPCR				U(0x20)
 #define SYSCFG_CMPENSETR			U(0x24)
+#define SYSCFG_CMPSD1CR				U(0x30)
+#define SYSCFG_CMPSD2CR				U(0x40)
 #define SYSCFG_IOSIZE				U(0x400)
 
 /*
@@ -32,6 +36,9 @@
 
 #define SYSCFG_CMPCR_READY_TIMEOUT_US		U(1000)
 
+#define CMPENSETR_OFFSET			U(0x4)
+#define CMPENCLRR_OFFSET			U(0x8)
+
 /*
  * SYSCFG_CMPENSETR Register
  */
@@ -44,60 +51,84 @@ static vaddr_t get_syscfg_base(void)
 	return io_pa_or_va(&base, SYSCFG_IOSIZE);
 }
 
-void stm32mp_syscfg_enable_io_compensation(void)
+static void enable_io_compensation(int cmpcr_offset)
 {
-	vaddr_t syscfg_base = get_syscfg_base();
-	uint64_t timeout_ref = 0;
-
-	if (clk_enable(stm32mp_rcc_clock_id_to_clk(CK_CSI)) ||
-	    clk_enable(stm32mp_rcc_clock_id_to_clk(SYSCFG)))
-		panic();
+	vaddr_t cmpcr_va = get_syscfg_base() + cmpcr_offset;
+	uint32_t value = 0;
 
-	io_setbits32(syscfg_base + SYSCFG_CMPENSETR, SYSCFG_CMPENSETR_MPU_EN);
+	if (io_read32(cmpcr_va) & SYSCFG_CMPCR_READY)
+		return;
 
-	timeout_ref = timeout_init_us(SYSCFG_CMPCR_READY_TIMEOUT_US);
+	io_setbits32(cmpcr_va + CMPENSETR_OFFSET, SYSCFG_CMPENSETR_MPU_EN);
 
-	while (!(io_read32(syscfg_base + SYSCFG_CMPCR) & SYSCFG_CMPCR_READY)) {
-		if (timeout_elapsed(timeout_ref)) {
-			EMSG("IO compensation cell not ready");
-			/* Allow an almost silent failure here */
-			break;
-		}
+	if (IO_READ32_POLL_TIMEOUT(cmpcr_va, value, value & SYSCFG_CMPCR_READY,
+				   0, SYSCFG_CMPCR_READY_TIMEOUT_US)) {
+		/* Allow an almost silent failure here */
+		EMSG("IO compensation cell not ready");
 	}
 
-	io_clrbits32(syscfg_base + SYSCFG_CMPCR, SYSCFG_CMPCR_SW_CTRL);
+	io_clrbits32(cmpcr_va, SYSCFG_CMPCR_SW_CTRL);
 
-	DMSG("SYSCFG.cmpcr = %#"PRIx32, io_read32(syscfg_base + SYSCFG_CMPCR));
+	DMSG("SYSCFG.cmpcr = %#"PRIx32, io_read32(cmpcr_va));
 }
 
-void stm32mp_syscfg_disable_io_compensation(void)
+static __maybe_unused void disable_io_compensation(int cmpcr_offset)
 {
-	vaddr_t syscfg_base = get_syscfg_base();
-	uint32_t value = 0;
-
-	value = io_read32(syscfg_base + SYSCFG_CMPCR) >>
-		SYSCFG_CMPCR_ANSRC_SHIFT;
+	vaddr_t cmpcr_base = get_syscfg_base() + cmpcr_offset;
+	uint32_t value_cmpcr = 0;
+	uint32_t apsrc_ansrc = 0;
+	uint32_t value_cmpcr2 = 0;
 
-	io_clrbits32(syscfg_base + SYSCFG_CMPCR,
-		     SYSCFG_CMPCR_RANSRC | SYSCFG_CMPCR_RAPSRC);
+	value_cmpcr = io_read32(cmpcr_base);
+	value_cmpcr2 = io_read32(cmpcr_base + CMPENSETR_OFFSET);
+	if (!(value_cmpcr & SYSCFG_CMPCR_READY &&
+	      value_cmpcr2 & SYSCFG_CMPENSETR_MPU_EN))
+		return;
 
-	value = io_read32(syscfg_base + SYSCFG_CMPCR) |
-		(value << SYSCFG_CMPCR_RANSRC_SHIFT);
+	/* Copy apsrc / ansrc in ransrc /rapsrc */
+	apsrc_ansrc = value_cmpcr >> SYSCFG_CMPCR_ANSRC_SHIFT;
+	value_cmpcr &= ~(SYSCFG_CMPCR_RANSRC | SYSCFG_CMPCR_RAPSRC);
+	value_cmpcr |= SHIFT_U32(apsrc_ansrc, SYSCFG_CMPCR_RANSRC_SHIFT);
 
-	io_write32(syscfg_base + SYSCFG_CMPCR, value | SYSCFG_CMPCR_SW_CTRL);
+	io_write32(cmpcr_base, value_cmpcr | SYSCFG_CMPCR_SW_CTRL);
 
-	DMSG("SYSCFG.cmpcr = %#"PRIx32, io_read32(syscfg_base + SYSCFG_CMPCR));
+	DMSG("SYSCFG.cmpcr = %#"PRIx32, io_read32(cmpcr_base));
 
-	io_clrbits32(syscfg_base + SYSCFG_CMPENSETR, SYSCFG_CMPENSETR_MPU_EN);
-
-	clk_disable(stm32mp_rcc_clock_id_to_clk(CK_CSI));
-	clk_disable(stm32mp_rcc_clock_id_to_clk(SYSCFG));
+	io_setbits32(cmpcr_base + CMPENCLRR_OFFSET, SYSCFG_CMPENSETR_MPU_EN);
 }
 
 static TEE_Result stm32mp1_iocomp(void)
 {
-	stm32mp_syscfg_enable_io_compensation();
+	if (clk_enable(stm32mp_rcc_clock_id_to_clk(CK_CSI)) ||
+	    clk_enable(stm32mp_rcc_clock_id_to_clk(SYSCFG)))
+		panic();
+
+	enable_io_compensation(SYSCFG_CMPCR);
 
 	return TEE_SUCCESS;
 }
+
 driver_init(stm32mp1_iocomp);
+
+#ifdef CFG_STM32MP13
+void stm32mp_set_vddsd_comp_state(enum stm32mp13_vddsd_comp_id id, bool enable)
+{
+	int cmpcr_offset = 0;
+
+	switch (id) {
+	case SYSCFG_IO_COMP_IDX_SD1:
+		cmpcr_offset = SYSCFG_CMPSD1CR;
+		break;
+	case SYSCFG_IO_COMP_IDX_SD2:
+		cmpcr_offset = SYSCFG_CMPSD2CR;
+		break;
+	default:
+		panic();
+	}
+
+	if (enable)
+		enable_io_compensation(cmpcr_offset);
+	else
+		disable_io_compensation(cmpcr_offset);
+}
+#endif /* CFG_STM32MP13 */
diff --git a/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.h b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.h
new file mode 100644
index 000000000..2d35da2eb
--- /dev/null
+++ b/core/arch/arm/plat-stm32mp1/drivers/stm32mp1_syscfg.h
@@ -0,0 +1,24 @@
+/* SPDX-License-Identifier: BSD-2-Clause */
+/*
+ * Copyright (c) 2023, STMicroelectronics
+ */
+
+#ifndef __DRIVERS_STM32MP1_SYSCFG_H
+#define __DRIVERS_STM32MP1_SYSCFG_H
+
+/* IO compensation domains IDs for STM32MP13 variants */
+enum stm32mp13_vddsd_comp_id {
+	SYSCFG_IO_COMP_IDX_SD1,
+	SYSCFG_IO_COMP_IDX_SD2,
+	SYSCFG_IO_COMP_COUNT
+};
+
+#ifdef CFG_STM32MP13
+/*
+ * Enable or disable IO compensation for a VDDSD IO domains
+ * @id: VDDSD domain ID
+ * @enable: True to enable IO compensation, false to disable
+ */
+void stm32mp_set_vddsd_comp_state(enum stm32mp13_vddsd_comp_id id, bool enable);
+#endif /*CFG_STM32MP13*/
+#endif /*__DRIVERS_STM32MP1_SYSCFG_H*/
diff --git a/core/arch/arm/plat-stm32mp1/stm32_util.h b/core/arch/arm/plat-stm32mp1/stm32_util.h
index c2ade22a2..627933e4a 100644
--- a/core/arch/arm/plat-stm32mp1/stm32_util.h
+++ b/core/arch/arm/plat-stm32mp1/stm32_util.h
@@ -18,13 +18,6 @@
 /* Backup registers and RAM utils */
 vaddr_t stm32mp_bkpreg(unsigned int idx);
 
-/*
- * SYSCFG IO compensation.
- * These functions assume non-secure world is suspended.
- */
-void stm32mp_syscfg_enable_io_compensation(void);
-void stm32mp_syscfg_disable_io_compensation(void);
-
 /* Platform util for the RCC drivers */
 vaddr_t stm32_rcc_base(void);
 
-- 
2.47.2

